Fix #1178

ilyaZar opened the pull request:
This PR fixes a regression in how `guess_where_config()` resolves custom config paths in `R/app_config.R`. Specifically, it now correctly handles nested expressions of the form:

```
file = Sys.getenv("CONFIG_PATH", app_sys("golem-config.yml"))
```

Previously, the logic relied on regex and failed to correctly extract the path from nested calls. This led to malformed paths and erroneous conflict detection between the user and default config files.

Also, when experimenting in package development there could be several `file = ...` calls, one being commented out, in the `app_config.R` i.e. :
```
  # file = app_sys("golem-config.yml")
  file = app_sys("golem-config.yml")
```
This is fixed as well, the new behavior identifies the correct (not commented) case only now.

### Changes made and To-Do

- [x] Parse RHS using AST
- [x] Handle `app_sys(...)` and `Sys.getenv(..., app_sys(...))`
- [x] Only evaluate `file =` and neglected commented lines i.e. `# file = ...`
- [ ] Add test case for malformed input
- [ ] Possibly handle more nesting from #1179


Fix:
#1178 
#1179

ilyaZar committed:
fix(config): support nested Sys.getenv(app_sys(...)) in user config detection

Parse RHS of file= argument in app_config.R using AST traversal to robustly
support expressions like Sys.getenv("CONFIG_PATH", app_sys("...")). Fixes bug
introduced post-0.4.1 where regex extraction failed on nested calls.

ilyaZar commented:
@ColinFay 

Related to this there is #1179, giving a more complicated expression to the argument `file` which is currently not handled. 

In general, I am not sure if one should support this type of behavior.

One could simply restrict the possible values to the `file` argument in the docs, as there is infinetely many nestings and possible function calls identifying the config `.yml` from the file argument. Two other ways come to my mind though:

1. `grep` for the .yaml config file directly, after recursively evaluating the complete AST associated with the right side of the `file` argument e.g. the right side is `app_sys("golem-config.yml")` in the full argument `file =  app_sys("golem-config.yml")`. One can recursively deparse down the AST until "golem-config.yml" or any other .yml is found.
2. evaluate the right side argument `app_sys("golem-config.yml")` from within, but that would require knowledge of `app_sys()` or any more complicated functions which is usually not availabe (i.e. `eval(parse(file))` will not work when `file` contains `app_sys()` as this is uknown during runtime).

Ready to implement other ideas.

Well

ilyaZar committed:
feat: ignore commented lines when detecting app_sys() in config parsing

Sometimes, there are commented lines in app_config.R for the file argument
to test different arguments for `file` in the `get_golem_config()`
function.

Ensures that commented-out config candidates like '# file = app_sys(...)'
are not parsed and mistakenly used for user config path resolution.
Improves robustness of multi-line detection in guess_lines_to_config_file().

ColinFay commented:
Thanks a lot @ilyaZar 🫶

I feel like the code for guess_where_config is getting a bit complex 🤔 and I'm not quite sure if we should continue supporting the "try and guess" behavior. Maybe we should provide an explicit way to change the config path using an env var, and stop trying to guess where it is — a large portion of the devs will not change the default path, so we can expect from devs that are playing with config files to be comfortable enough with environment variable.

So, my proposition is : 

- golem has `golem::get_current_config()` that calls under the hood `guess_where_config()`
- most devs will not be playing with the default config file, so we can safely rely on inst/golem-config.yml path for being the default in most cases. 
- power users will change the config path, but we could expect from them to be able to do it with an env var, or to hard code the path to the file in 

I suggest modifying the `guess_where_config()` function that way:

```r
guess_where_config <- function(
  path = golem::pkg_path(),
  file = "inst/golem-config.yml"
) {
  if (Sys.getenv("GOLEM_CONFIG_PATH") == "") {
    path_to_config <- fs_path(
      path,
      file
    )
  } else {
    path_to_config <- Sys.getenv("GOLEM_CONFIG_PATH")
  }
  path_to_config <- fs_path_abs(path_to_config)
  if (!fs_file_exists(path_to_config)) {
    stop(
      "Unable to locate config file. Either use the default path at inst/golem-config.yml or set it with a GOLEM_CONFIG_PATH environment variable"
    )
  }
}
```

Then, in the R/app_config.R file, switching to: 

```r
get_golem_config <- function(
  value,
  config = Sys.getenv(
    "GOLEM_CONFIG_ACTIVE",
    Sys.getenv(
      "R_CONFIG_ACTIVE",
      "default"
    )
  ),
  use_parent = TRUE,
  # If you don't want to use the default config file:
  # - replace the function call 
  #   and write the path to your config file
  # - set a `GOLEM_CONFIG_PATH` env var that will
  #   be picked by golem::get_current_config()
  file = golem::get_current_config()
) {
  config::get(
    value = value,
    config = config,
    file = file,
    use_parent = use_parent
  )
}
```

So basically, we stop parsing the files from the package to get the config file path, we either rely on the default value, an env var, or an hard coded path.

@LDSamson @ilyaZar, what do you both think of this approach?

ColinFay commented:
@LDSamson thanks for your feedback. 

I also think this would make using different config file name or path easier — mosts user will use the default and advanced user can configure with their needs, and yes you're right, it would no longer parse `app_config.R` so you can do whatever you want here :)

ilyaZar commented:
Hi @ColinFay 

yes definitely agreed, and as said above I was never a big of the "try and guess" behavior ... although I contributed to exactly that code :) 

However, I have a feeling that the order needs a change.

### Ideas for slight change

1. _**First**_ , when `GOLEM_CONFIG_PATH` is there and points to an existing file -> hard `stop()` if the file is not found because dev-users need an early error e.g. when they have made a typo or moved a file.

2. Then, if there is no `GOLEM_CONFIG_PATH` envir var, we have:
  2.A a "standard" user with default path
  2.B a dev user who prefers to set paths directly without envir-vars or made a typo like `GOLEM_CONFIG_PATHHHHH`

```r
guess_where_config <- function(
  path = golem::pkg_path(),
  file = "inst/golem-config.yml"
) {
  # 1. DEV USER: envir var is set, and if the file does not exist hard stop
  if (Sys.getenv("GOLEM_CONFIG_PATH") != "") {
    path_to_config <- fs_path_abs(Sys.getenv("GOLEM_CONFIG_PATH"))
    if (!fs_file_exists(path_to_config)) {
      msg_err <- paste0(
        "Unable to locate a config file using the environment variable
        'GOLEM_CONFIG_PATH'. Check for typos in the (path to the) filename.")
      stop(msg_err)
    }
  } else if (Sys.getenv("GOLEM_CONFIG_PATH") == "") {
    path_to_config <- fs_path(
      path,
      file
    )
    # 2.A standard user: sets default path -> all is fine
    CHECK_DEFAULT_PATH <- grepl("*./inst/golem-config.yml$", path_to_config)
    if (isFALSE(CHECK_DEFAULT_PATH)) {
    # 2.B DEV USER: sets non-default path OR has type in the envir variable name: 
    # if the file does not exist hard stop
      if (!fs_file_exists(path_to_config)) {
        msg_err <- paste0(
          "Unable to locate a config file from either the string to path, or",
          "the 'GOLEM_CONFIG_PATH' environment variable. Check for typos."
        )
        stop(msg_err)
      }
    }
  }
  return(path_to_config)
}
```


### Benefits

1. There can be cases where the user has a correct `GOLEM_CONFIG_PATH` variable, but it points to a non-existent file: this is handled in case 1.

2. There can be cases where the user has a type like `GOLEM_CONFIG_PATHHHHH`, or, an incorrect argument string to a non-existent path and that gets a separate error message for case 2.B

3. If a standard user does not change anything and the file exists, all works. If a standard user has screwed up the `.yml` config, the old dance of `get_current_config()` is executed 
```r
if (!fs_file_exists(path_conf)) {
    if (rlang_is_interactive()) {
      ask <- ask_golem_creation_upon_config(path_conf)
      # Return early if the user doesn't allow
      if (!ask) {
        return(NULL)
      }
      fs_file_copy(
        path = golem_sys("shinyexample/inst/golem-config.yml"),
        new_path = fs_path(
          path,
          "inst/golem-config.yml"
        )
      )
      fs_file_copy(
        path = golem_sys("shinyexample/R/app_config.R"),
        new_path = fs_path(
          path,
          "R/app_config.R"
        )
      )
```
The old main behavior still works and missing files get replaced from the `shinyexample`.


Just some ideas, ... ready to make other changes !

ilyaZar committed:
feat: get rid of file guessing behavior to make implementation easy

- implements different cases with different error messages

- error messages inform mostly about what case exactly went wrong

ilyaZar committed:
chore: adjust app_config.R to changes

ilyaZar committed:
tests: improve test coverage for config.R to 100 %

ilyaZar commented:
### Changes made and To-Do

- [x] implement proposed `guess_where_config()` behavior: fae9c3af3cec82101716a1bca9bff66632707fd5
- [x] adjust `get_current_config()` for new `guess_where_config()`: df3b755a968df9235a75dd65b9c7903e6d4387b1
- [x] update tests and increase coverage to 100 %: da065b013ff2f758233a2623c4ac7c6cf8698fa2
- [ ] update docs in `get_current_config`
- [ ] update docs in [Engineering Production-Grade Shiny Apps](https://engineering-shiny.org/golem.html?q=config#golem-config)

ilyaZar commented:
I encountered some issues.

Actually, the `get_current_config` is used numerous times inside `{golem}`, including inside `get_golem_things()`, so whenever I start a fresh golem skeleton and try to use a fixed path `./inst/golem-config2.yml` like in `app_config.R` it will not find the file and start the `get_current_config` dance prompting me to add a new config file via `ask_golem_creation_upon_config(path_conf)`. 

Here is what is not working:

```r
get_golem_config <- function(
  value,
  config = Sys.getenv(
    "GOLEM_CONFIG_ACTIVE",
    Sys.getenv(
      "R_CONFIG_ACTIVE",
      "default"
    )
  ),
  use_parent = TRUE,
  # If you don't want to use the default config file:
  # - replace the function call
  #   and write the path to your config file
  # - set a `GOLEM_CONFIG_PATH` env var that will
  #   be picked by golem::get_current_config()
  # file = golem::get_current_config()
  file = "./inst/golem-config2.yml"
) {
  config::get(
    value = value,
    config = config,
    file = file,
    use_parent = use_parent
  )
}
```

So changing to hard coded path at the beginning of a new golem-skeleton fails. Probably also `amend_golem_config` and `set_golem_options` might fail unexpectedly....

@ColinFay any ideas ? I may try to rework those `get_golem_things()` / `amend_golem_config()` as well but it might be a wiser decision to think a bit longer here , as now it's touching quite some other internal implemenaton.

LDSamson commented:
Just some brainstorming ideas below. 

What if you would create a file in the `inst` folder in which you can write the dedicated file path, and nothing else? Optionally you can create a helper `set_config_path()` to set it properly. It would be much safer to read than the entire `app_config.R` file. It is similar to how `renv` discovers the active profile: it uses a single file named `profile` that _only_ stores the active profile's name. 

Then the logic in `guess_where_config` would check for a valid path in this order: 
- The `GOLEM_CONFIG_PATH` environment variable 
- The file named `golem_config_path` in the `inst` folder (use `readLines` or an equivalent function)

If the file is not available the `envvar` is not set, fall back to the standard path.

Furthermore, I am probably missing something here: is there a reason for exporting the entire `get_golem_config()` function apart from giving end users the possibility to change the location of the `golem-config.yml` file? I understand that would be a big change, but I am just curious.

ilyaZar commented:
@LDSamson , yes this is a good idea and all doable

I do not have any hard opinion about the correct way.

The underlying problem to address is always the same:

- `{golem}` is used in different projects as a helper package to build other pkgs

- there has to be conventions about where to store relevant files for `{golem}` -- such as the config-file -- on each user side project

Your suggestion (if I understand correctly), makes the implementation inside `{golem}` easier for the hard-coded path version: 

1. hard coded paths are set in a file, not via a function argument in `app_config.R`
2. `app_config.R` can be altered almost arbitrarily because the underlying logic of `{golem}` to find config files no longer relates to `app_config.R`

Instead, config-files are retrieved either as envir-vars or  a separate `golem_config_path` file (and the path stored as a string therein)

Sounds liike a good way to me, it would clean up some exotic internal implementations as well, or @ColinFay ?

ilyaZar commented:
Regarding `get_golem_config`: this is kind of the same initial problem

- as an end user of golem you usually not alter the `{golem}` package

- since the `{golem}` needs a pre-defined way to work with configs, the user has to adhere to the `{golem}` conventions to set configs anyway

Yes, it’s even possible to “purge” `app_config.R` along with `get_golem_config()`, and rename the file  to `app_file_access.R` and only keep `app_sys()` there (it is more general, finding arbitrary files)

ColinFay commented:
This has been rebased and integrated in #1210
